sklearn/cluster/_hdbscan/_reachability.pyx

```diff
@@ -48,78 +58,125 @@ def mutual_reachability(distance_matrix, min_points=5, max_dist=0.0):
     core_distances = np.partition(
-        distance_matrix,
-        min_points,
-        axis=0,
-    )[min_points]
+        distance_matrix, further_neighbor_idx, axis=1
+    )[:, further_neighbor_idx]
 
     with nogil:
         for i in range(n_samples):
             for j in prange(n_samples):
-                mr_dist = max(
+                mutual_reachibility_distance = max(
                     core_distances[i],
                     core_distances[j],
-                    distance_matrix[i, j]
+                    distance_matrix[i, j],
                 )
-                distance_matrix[i, j] = mr_dist
-    return distance_matrix
+                distance_matrix[i, j] = mutual_reachibility_distance
 
-# Assumes LIL format.
-# TODO: Rewrite for CSR.
-cdef _sparse_mutual_reachability(
-    object distance_matrix,
-    cnp.intp_t min_points=5,
-    cnp.float64_t max_dist=0.
+
+def _sparse_mutual_reachability_graph(
+    cnp.ndarray[floating, ndim=1, mode="c"] data,
+    cnp.ndarray[integral, ndim=1, mode="c"] indices,
+    cnp.ndarray[integral, ndim=1, mode="c"] indptr,
+    cnp.intp_t n_samples,
+    cnp.intp_t further_neighbor_idx,
+    floating max_distance,
 ):
-    cdef cnp.intp_t i, j, n, n_samples = distance_matrix.shape[0]
-    cdef cnp.float64_t mr_dist
-    cdef cnp.float64_t[:] core_distances
-    cdef cnp.int32_t[:] nz_row_data, nz_col_data
-    core_distances = np.empty(n_samples, dtype=np.float64)
+    """Sparse implementation of mutual reachability graph.
+
+    The computation is done in-place, i.e. the distance matrix is modified
+    directly. This implementation only accepts `CSR` format sparse matrices.
+
+    Parameters
+    ----------
+    distance_matrix : sparse matrix of shape (n_samples, n_samples)
+        Sparse matrix of distances between samples. The sparse format should
+        be `CSR`.
+
+    further_neighbor_idx : int
+        The index of the furthest neighbor to use to define the core distances.
+
+    max_distance : float
+        The distance which `np.inf` is replaced with. When the true mutual-
+        reachability distance is measured to be infinite, it is instead
+        truncated to `max_dist`. Only used when `distance_matrix` is a sparse
+        matrix.
+    """
+    cdef:
+        integral i, col_ind, row_ind
+        floating mutual_reachibility_distance
+        floating[:] core_distances
+        floating[:] row_data
+
+    if floating is float:
+        dtype = np.float32
+    else:
+        dtype = np.float64
+
+    core_distances = np.empty(n_samples, dtype=dtype)
 
     for i in range(n_samples):
-        if min_points < len(distance_matrix.data[i]):
+        row_data = data[indptr[i]:indptr[i + 1]]
+        if further_neighbor_idx < row_data.size:
             core_distances[i] = np.partition(
-                distance_matrix.data[i],
-                min_points
-            )[min_points]
+                row_data, further_neighbor_idx
+            )[further_neighbor_idx]
```

Comment:
We should be able to use [`std::nth_partition`](https://en.cppreference.com/w/cpp/algorithm/nth_element) in a `nogil` context instead.

There's some plumbing in scikit-learn to use this interface via Cython:

https://github.com/scikit-learn/scikit-learn/blob/f7239bf1a19b93fc0bafe292700bb88b3fa54fac/sklearn/neighbors/_partition_nodes.pyx#L4-L62

To me, this better be explored in another PR. What do you think?

Reply:
Indeed that's a good suggestion for a follow-up PR. Thanks for sharing this!

Reply:
Is there an issue somewhere with the list of the follow-up work for HDBSCAN?

Reply:
> Is there an issue somewhere with the list of the follow-up work for HDBSCAN?

I'm currently tracking that in the same issue as the tasks for primary inclusion of HDBSCAN: https://github.com/scikit-learn/scikit-learn/issues/24686